When building XGBoost 2.0+ with GPU support from source get fatal "expression did not evaluate to a constant" error

As mentioned in #7024, I continue to have issues with VS 2022, 2019, and 2017, when building XGBoost from source for Windows and R with GPU support. The error comes from line 298 in [histogram.cu](https://github.com/dmlc/xgboost/blob/master/src/tree/gpu_hist/histogram.cu) which reads: `constexpr int kMinItemsPerBlock = kItemsPerTile;` I am on Windows 10 64bit and CUDA 11.7, and I have gotten this error using VS 2017, 2019, and 2022. Obviously, XGBoost does not finish compiling. When building for CPU only, it works fine.

The full error log is below 

Thank you

```
D:\xgboost\src\tree\gpu_hist\histogram.cu(315): note: see reference to function template instantiation 'auto xgboost::tree::BuildGradientHistogram::<lambda_6d0a0949a274b91398442d30e87cb3ab>::operator ()<void(__cdecl *)(x
gboost::EllpackDeviceAccessor,xgboost::tree::FeatureGroupsAccessor,xgboost::common::Span<const uint32_t,18446744073709551615>,GradientSumT *__restrict ,const xgboost::GradientPair *__restrict ,xgboost::tree::HistRounding
)>(void (__cdecl )(xgboost::EllpackDeviceAccessor,xgboost::tree::FeatureGroupsAccessor,xgboost::common::Span<const uint32_t,18446744073709551615>,GradientSumT _restrict ,const xgboost::GradientPair *re
strict ,xgboost::tree::HistRounding)) const' being compiled
with
[
GradientSumT=xgboost::detail::GradientPairInternal
]
D:\xgboost\src\tree\gpu_hist\histogram.cu(323): note: see reference to function template instantiation 'void xgboost::tree::BuildGradientHistogram<xgboost::detail::GradientPairInternal>(const xgboost::EllpackDevi
ceAccessor &,const xgboost::tree::FeatureGroupsAccessor &,xgboost::common::Span<const xgboost::detail::GradientPairInternal,18446744073709551615>,xgboost::common::Span<const uint32_t,18446744073709551615>,xgboost:
:common::Span<xgboost::detail::GradientPairInternal,18446744073709551615>,xgboost::tree::HistRounding<xgboost::detail::GradientPairInternal>,bool)' being compiled

D:\xgboost\src\tree\gpu_hist\histogram.cu(298): error C2131: expression did not evaluate to a constant [D:\xgboost\buildgpu\src\objxgboost.vcxproj]

D:\xgboost\src\tree\gpu_hist\histogram.cu(298): note: failure was caused by a read of a variable outside its lifetime
D:\xgboost\src\tree\gpu_hist\histogram.cu(298): note: see usage of 'this'
D:\xgboost\src\tree\gpu_hist\histogram.cu(299): warning C4267: 'initializing': conversion from 'size_t' to 'int', possible loss of data [D:\xgboost\buildgpu\src\objxgboost.vcxproj]

**C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\IDE\VC\VCTargets\BuildCustomizations\CUDA 11.7.targets(790,9): error MSB3721: The command ""C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.7
bin\nvcc.exe" -gencode=arch=compute_86,code="sm_86,compute_86" -gencode=arch=compute_86,code="compute_86,compute_86" -gencode=arch=compute_86,code="sm_86,compute_86" --use-local-env -ccbin "C:\Program Files (x86)\Mic
rosoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64" -x cu -ID:\xgboost\include -I"D:\xgboost\dmlc-core\include" -ID:\xgboost\rabit\include -ID:\xgboost\gputreeshap -I"D:\xgboost\buildgpu\dml
c-core\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.7\include" --keep-dir x64\Release -maxrregcount=0 --machine 64 --compile -cudart static --expt-extended-lambda --expt-relaxed-constexpr -Xfatb
in=-compress-all -lineinfo -std=c++14 -Xcompiler="/EHsc -Ob2 -openmp /utf-8" -D_WINDOWS -DNDEBUG -D_CRT_SECURE_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -DXGBOOST_USE_CUDA=1 -DDMLC_LOG_CUSTOMIZE=1 -DXGBOOST_MM_PREFETCH_PRES
ENT=1 -D__USE_XOPEN2K8 -DDMLC_CORE_USE_CMAKE -DDMLC_USE_CXX11=1 -DXGBOOST_STRICT_R_MODE=1 -DXGBOOST_CUSTOMIZE_GLOBAL_PRNG=1 -DDMLC_LOG_BEFORE_THROW=0 -DDMLC_DISABLE_STDIN=1 -DRABIT_CUSTOMIZE_MSG -DRABIT_STRICT_CXX98 -D"C
MAKE_INTDIR="Release"" -D_MBCS -DWIN32 -D_WINDOWS -DNDEBUG -D_CRT_SECURE_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -DXGBOOST_USE_CUDA=1 -DDMLC_LOG_CUSTOMIZE=1 -DXGBOOST_MM_PREFETCH_PRESENT=1 -D__USE_XOPEN2K8 -DDMLC_CORE_USE
CMAKE -DDMLC_USE_CXX11=1 -DXGBOOST_STRICT_R_MODE=1 -DXGBOOST_CUSTOMIZE_GLOBAL_PRNG=1 -DDMLC_LOG_BEFORE_THROW=0 -DDMLC_DISABLE_STDIN=1 -DRABIT_CUSTOMIZE_MSG -DRABIT_STRICT_CXX98 -D"CMAKE_INTDIR="Release"" -Xcompiler "/
EHsc /W3 /nologo /O2 /Fdobjxgboost.dir\Release\objxgboost.pdb /FS /MT /GR" -o objxgboost.dir\Release\histogram.obj "D:\xgboost\src\tree\gpu_hist\histogram.cu"" exited with code 2. [D:\xgboost\buildgpu\src\objxgboost.vcxp
roj]
```

--- a/src/tree/gpu_hist/histogram.cu
+++ b/src/tree/gpu_hist/histogram.cu
@@ -275,7 +275,7 @@ void BuildGradientHistogram(EllpackDeviceAccessor const& matrix,
   constexpr int kItemsPerThread = 8;
   constexpr int kItemsPerTile = kBlockThreads * kItemsPerThread;
 
-  auto runit = [&](auto kernel) {
+  auto runit = [&, kMinItemsPerBlock=kItemsPerTile](auto kernel) {
     if (shared) {
       dh::safe_cuda(cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize,
                                          max_shared_memory));
@@ -295,7 +295,6 @@ void BuildGradientHistogram(EllpackDeviceAccessor const& matrix,
     // Otherwise launch blocks such that each block has a minimum amount of work to do
     // There are fixed costs to launching each block, e.g. zeroing shared memory
     // The below amount of minimum work was found by experimentation
-    constexpr int kMinItemsPerBlock = kItemsPerTile;
     int columns_per_group = common::DivRoundUp(matrix.row_stride, feature_groups.NumGroups());
     // Average number of matrix elements processed by each group
     std::size_t items_per_group = d_ridx.size() * columns_per_group;

--- a/src/tree/gpu_hist/histogram.cu
+++ b/src/tree/gpu_hist/histogram.cu
@@ -275,7 +275,7 @@ void BuildGradientHistogram(EllpackDeviceAccessor const& matrix,
   constexpr int kItemsPerThread = 8;
   constexpr int kItemsPerTile = kBlockThreads * kItemsPerThread;
 
-  auto runit = [&, kMinItemsPerBlock=kItemsPerTile](auto kernel) {
+  auto runit = [&, kMinItemsPerBlock = kItemsPerTile](auto kernel) {
     if (shared) {
       dh::safe_cuda(cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize,
                                          max_shared_memory));

